content/automate/ManageIQ/Infrastructure/VM/Provisioning/Placement.class/__methods__/vmware_best_fit_least_utilized.rb

```diff
@@ -1,37 +1,74 @@
 #
 # Description: This method is used to find all hosts, datastores that are the least utilized
 #
-$evm.log("info", "Args:    #{MIQ_ARGS.inspect}")
-
-# Get variables
-prov = $evm.root["miq_provision"]
-vm = prov.vm_template
-raise "VM not specified" if vm.nil?
-ems = vm.ext_management_system
-raise "EMS not found for VM [#{vm.name}]" if ems.nil?
-
-# Log space required
-$evm.log("info", "vm=[#{vm.name}], space required=[#{vm.provisioned_storage}]")
-
-host = storage = nil
-min_registered_vms = nil
-prov.eligible_hosts.each do |h|
-  next if h.maintenance
-  next unless h.power_state == "on"
-  nvms = h.vms.length
-  if min_registered_vms.nil? || nvms < min_registered_vms
-    storages = h.writable_storages.find_all { |s| s.free_space > vm.provisioned_storage } # Filter out storages that do not have enough free space for the Vm
-    s = storages.max_by(&:free_space)
-    unless s.nil?
-      host    = h
-      storage = s
-      min_registered_vms = nvms
+module ManageIQ
+  module Automate
+    module Infrastructure
+      module VM
+        module Provisioning
+          module Placement
+            class VmwareBestFitLeastUtilized
+              def initialize(handle = $evm)
+                @handle = handle
+              end
+
+              def log_and_raise(message)
+                @handle.log(:error, message)
+                raise "ERROR - #{message}"
+              end
+
+              def main
+                @handle.log("info", "vm=[#{vm.name}], space required=[#{vm.provisioned_storage}]")
+                best_fit_least_utilized
+              end
+
+              private
+
+              def request
+                @request ||= @handle.root["miq_provision"].tap do |req|
```

Comment:
@lfu 
You can have a spec that doesn't set the miq_provision attribute in the root object and call the method and it should raise an exception
Add another spec which has a request but no vm and you will be able to test the second log_and_raise